python/test_dcr_template.py

```diff
@@ -305,22 +296,55 @@ def test_dataId_list(self):
-        input_arr = []
-        for model in self.dcrModel.model:
-            input_vals = model[j - radius: j + radius + 1, i - radius: i + radius + 1] / weight_scale
-            input_vals[radius, radius] = 0.
-            input_arr.append(np.ravel(input_vals))
-        self.assertFloatsAlmostEqual(np.hstack(input_arr), model_vals)
+    def test_model_dataId(self):
+        subfilter = 1
+        band_ref = 'g'
+        ref_id = {'filter': band_ref, 'tract': 0, 'patch': '0', 'subfilter': subfilter}
+        dataId = DcrModel._build_model_dataId(band_ref, subfilter=subfilter)
+        self.assertEqual(ref_id, dataId)
+
+    def test_fetch_metadata(self):
+        az = DcrModel._fetch_metadata(self.exposure.getMetadata(), "AZIMUTH")
+        self.assertFloatsAlmostEqual(az, self.azimuth.asDegrees())
+
+    def test_fetch_missing_metadata(self):
+        bogus_ref = 3.7
+        bogus = DcrModel._fetch_metadata(self.exposure.getMetadata(), "BOGUS", default_value=bogus_ref)
+        self.assertFloatsAlmostEqual(bogus, bogus_ref)
+
+    def test_generate_template(self):
+        """Compare the result of generate_templates_from_model to previously computed values."""
+        data_file = "test_data/template.npy"
+        elevation_arr = np.radians([50., 70., 85.])
+        az = Angle(0.)
+        # Note that self.array is randomly generated each call. That's okay, because the template should
+        # depend only on the metadata.
+        exposures = [self.dcrModel.create_exposure(self.array, variance=None, elevation=Angle(el), azimuth=az)
+                     for el in elevation_arr]
+        model_gen = self.dcrModel.generate_templates_from_model(exposures=exposures)
+        model_test = [model for model in model_gen]
+        # np.save(data_file, model_test)
+        model_ref = np.load(data_file)
+        for m_i in range(len(model_test)):
+            m_test = model_test[m_i].getMaskedImage().getImage().getArray()
+            m_ref = model_ref[m_i].getMaskedImage().getImage().getArray()
+            self.assertFloatsAlmostEqual(m_test, m_ref)
+
+    def test_warp_exposure(self):
+        wcs = self.exposure.getWcs()
+        bbox = self.exposure.getBBox()
+        wrap_warpExposure(self.exposure, wcs, bbox)
+        array_warped = self.exposure.getMaskedImage().getImage().getArray()
+        # For some reason the edges are all NAN.
+        valid_inds = np.isfinite(array_warped)
+        self.assertGreater(np.sum(valid_inds), (self.size/2)**2)
+        array_ref = self.array[valid_inds]
+        array_warped = array_warped[valid_inds]
+        self.assertFloatsAlmostEqual(array_ref, array_warped, rtol=1e-7)
+
+    def test_rotation_angle(self):
+        rotation_angle = calculate_rotation_angle(self.exposure)
+        self.assertFloatsAlmostEqual(self.rotation_angle.asDegrees(), rotation_angle.asDegrees())
 
 
 class PersistanceTestCase(DcrModelTestBase, lsst.utils.tests.TestCase):
```

Comment:
Also your code comment below "The type "dcrModel" is read in as a 32 bit float..." does not seem to match the code. You are testing `dcrModel.model` as a 32-bit float and I'm not sure why because it is supposed to be a list of numpy arrays. I hope you can actually test the round tripping of the arrays, but if not, please explain in detail why not, and why what you are testing makes sense as a substitute.